### PR TITLE
[MEAR-153] skinnyModules option

### PR DESCRIPTION
**Pull request by mabrarov**
Changes:

* [MEAR-153] - Support of skinnyModules option for WAR, SAR, HAR, RAR and WSR EAR modules.
* [MEAR-153] - Configurable library directory for each EAR module overriding default value which depends on EAR module type.
* [MEAR-153] - Documentation for skinny modules and libDirectory property of EAR modules.

There are EAR module types (like AOP and ESB) which:

* are supported by [JBoss Packaging Maven Plugin](https://github.com/mojohaus/jboss-packaging-maven-plugin)
* can contain packaged JARs
* are not supported by Maven EAR Plugin 3.1.0

I don't think that MEAR-153 covers these EAR module types, so I have no plans to implement support of skinnyModules options for these types of EAR modules.

Support of WSR module type is implemented as part of RAR support (WsrModule extends RarModule). I cannot test it yet, because I have no information about WSR packaging / EAR module type.

### CODE REVIEW

**Comment by mabrarov** [comment]
Maybe we need to revisit approach for MEAR-153 and:

* keep using skinnyWars option just for WARs
* introduce skinnyModules option with `false` default value to handle all supported types of modules (including WARs)

That way we could stay backward compatible.

**Comment by mabrarov** [comment]
Support of RAR is implemented. Support of WSR is implemented as part of RAR support. I cannot test it yet, because I still have no information about WSR packaging / module type.

**Comment by mabrarov** [comment]
We need smbd to test skinny RAR, HAR and SAR - I have no code with this kind of EAR modules to ensure that skinny RAR, HAR and SAR works correctly, i.e. are deployed with the same success state as non-skinny.

**Comment by mabrarov** [comment]
@elharo, I do not plan any further work on or testing of these changes. Could you please trigger Jenkins build and review this pull request? Please note that these changes also fix backward incompatibility introduced by MEAR-267 (by pull request #19 )

**Comment by mabrarov** [comment]
We may need to update documentation of Maven EAR Plugin with detail description of skinnyModules option, its connection with skinnyWars option and with examples for skinnyModules option and libDirectory property of EAR module.

File: src/test/resources/projects/project-094/ear/pom.xml [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    </dependency>
+    <dependency>
+      <groupId>eartest</groupId>
+      <artifactId>sar-sample-two</artifactId>
+      <type>sar</type>
+      <!--
+        By some reason dependencies of SAR come into EAR as transitive dependencies while they should not,
+        because SAR contains its dependencies packaged as libraries (like WAR).
+        Looks like a bug of SAR packaging, i.e. bug of JBoss Packaging Maven Plugin where
+        includesDependencies should be explicitly set to true for
```
Comment (mabrarov) [inline src/test/resources/projects/project-094/ear/pom.xml:49]: Refer to [that comment](https://github.com/mojohaus/jboss-packaging-maven-plugin/commit/bf85249a7514aff3729c862faf7bcfed274a522b#r43279570) in [jboss-packaging-maven-plugin](https://github.com/mojohaus/jboss-packaging-maven-plugin) GitHub project.

**Comment by mabrarov** [comment]
@elharo, could you please trigger Jenkins build for this pull request?

**Review by elharo — COMMENTED** [review]
OK, I guess I was just looking at the last commit. 

https://ci-builds.apache.org/job/Maven/job/maven-box/job/maven-ear-plugin/job/mear-153/

File: src/main/java/org/apache/maven/plugins/ear/AbstractEarModule.java [hunk 2]
```diff
@@ -91,6 +91,14 @@
 
     private String moduleId;
 
+    /**
+     * Directory of module which contains libraries packaged into module. {@code null} value means that module
+     * doesn't contain any library. Each module type can provide default value for this directory and this option
+     * can be used to override that default value. If module libraries are located at the root of module then use
+     * single slash (/) to configure that in POM, i.e. single slash is treated as empty string.
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/ear/AbstractEarModule.java:98]: single slash (/) to configure that in POM. That is, a single slash is treated as an empty string.

File: src/main/java/org/apache/maven/plugins/ear/AbstractEarModule.java [hunk 3]
```diff
@@ -360,34 +374,34 @@ public String toString()
     }
 
     /**
-     * Cleans the bundle directory so that it might be used properly.
+     * Cleans the path pointing to resource inside archive so that it might be used properly.
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/ear/AbstractEarModule.java:377]: the resource inside the archive

File: src/main/java/org/apache/maven/plugins/ear/EarModule.java [hunk 4]
```diff
@@ -114,7 +114,8 @@ void resolveArtifact( Set<Artifact> artifacts )
     boolean changeManifestClasspath();
 
     /**
-     * @return The libDir.
+     * @return The directory of module which contains JAR libraries packaged within module. Can be {@code null},
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/ear/EarModule.java:117]: the directory of the module which contains the JAR files packages within the module

File: src/main/java/org/apache/maven/plugins/ear/EarMojo.java [hunk 5]
```diff
@@ -211,6 +211,18 @@
     @Parameter( defaultValue = "false" )
     private boolean skinnyWars;
 
+    /**
+     * Whether to create skinny EAR modules or not. A skinny EAR module is a WAR, SAR, HAR, RAR or WSR module that
+     * does not have all of its dependencies in it. Instead those dependencies are shared between the WARs, SARs, HARs,
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/ear/EarMojo.java:216]: "have all of its dependencies in it" --> contain all of its dependencies.

File: src/main/java/org/apache/maven/plugins/ear/EarMojo.java [hunk 6]
```diff
@@ -211,6 +211,18 @@
     @Parameter( defaultValue = "false" )
     private boolean skinnyWars;
 
+    /**
+     * Whether to create skinny EAR modules or not. A skinny EAR module is a WAR, SAR, HAR, RAR or WSR module that
+     * does not have all of its dependencies in it. Instead those dependencies are shared between the WARs, SARs, HARs,
+     * RARs and WSRs through the EAR. This option takes precedence over {@link #skinnyWars} option, i.e. if
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/ear/EarMojo.java:217]: over the
. i.e. --> . That is,

File: src/main/java/org/apache/maven/plugins/ear/EarMojo.java [hunk 7]
```diff
@@ -739,6 +751,11 @@ public boolean isNonFilteredExtension( String fileName )
     private void changeManifestClasspath( EarModule module, File original, JavaEEVersion javaEEVersion )
         throws MojoFailureException
     {
+        final String moduleLibDir = module.getLibDir();
+        if ( !( moduleLibDir == null || skinnyModules || skinnyWars && module instanceof WebModule ) )
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/ear/EarMojo.java:755]: Please add some extra parentheses for those of us who don't immediately the exact precedence of || and &&

File: src/main/java/org/apache/maven/plugins/ear/EarMojo.java [hunk 8]
```diff
@@ -799,16 +816,16 @@ private void changeManifestClasspath( EarModule module, File original, JavaEEVer
                 classPath = new Attribute( "Class-Path", "" );
             }
 
-            // Remove JAR modules
-            for ( JarModule jm : getAllJarModules() )
+            if ( moduleLibDir != null && ( skinnyModules || skinnyWars && module instanceof WebModule ) )
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/ear/EarMojo.java:819]: more parentheses please

File: src/site/apt/examples/skinny-modules.apt.vm [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+
+  * {{{../modules.html#rarModule}RAR}}
+
+  * {{{../modules.html#wsrModule}WSR}}
+
+  * {{{../modules.html#sarModule}SAR}}
+
+  * {{{../modules.html#harModule}HAR}}
+
+ Starting with version 3.1.1 Maven EAR Plugin supports referencing
```
Comment (elharo) [inline src/site/apt/examples/skinny-modules.apt.vm:44]: the Maven EAR

Comment (elharo) [inline src/site/apt/examples/skinny-modules.apt.vm:44]: 3.1.1., the

File: src/site/apt/examples/skinny-modules.apt.vm [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+
+ We need to change the EAR project's <<<pom.xml>>> to package those JARs
+ in the EAR and to use the
+ {{{../ear-mojo.html#skinnyModules}skinnyModules}} parameter.
+
+ <<Note:>> In this example we package all JARs into a <<<libs/>>> directory
+ within the EAR. This is just to distinguish between Java EE modules
+ (which will be packaged in the root of the EAR) and Java libraries
+ (which are packaged in <<<libs/>>>). Also, we use non standard
+ <<<outputFileNameMapping>>> just to shorten names of files packaged
```
Comment (elharo) [inline src/site/apt/examples/skinny-modules.apt.vm:56]: delete "just"

File: src/site/apt/examples/skinny-modules.apt.vm [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+ |-- META-INF
+ |   `-- application.xml
+ |-- libs
+ |   `-- shared-jar-1.0.0.jar
+ |-- war-1.0.0.war
+ |-- sar-1.0.0.sar
+ `-- rar-1.0.0.rar
++-----------------+
+
+ If you look inside the copies of <<<war-1.0.0.war>>>, <<<rar-1.0.0.ear>>>
```
Comment (elharo) [inline src/site/apt/examples/skinny-modules.apt.vm:121]: This is good. In general, stick to third person or second if necessary. That is, I suggest rewriting the "We" in these docs as "You".

File: src/site/apt/examples/skinny-modules.apt.vm [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+ <<<shared-jar-1.0.0.jar>>> and <<<lib/shared-jar-1.0.0.jar>>> files respectively.
+
+ Also, if you inspect the <<<MANIFEST.MF>>> of WAR, SAR and HAR modules of EAR
+ you will notice that the <<<Class-Path>>> entry has been modified or added,
+ if it was missing, and now has a reference to <<<libs/shared-jar-1.0.0.jar>>>.
+
+ If archive representing EAR module has non standard location of libraries then
+ this location can be configured using <<<libDirectory>>> property.
+
+ Here is example for SAR which contains libraries at the root of archive
```
Comment (elharo) [inline src/site/apt/examples/skinny-modules.apt.vm:133]: the archive. (Refer

File: src/site/apt/examples/skinny-modules.apt.vm [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+ If you look inside the copies of <<<war-1.0.0.war>>>, <<<rar-1.0.0.ear>>>
+ and <<<sar-1.0.0.sar>>>, that are packaged within the EAR, you will see that
+ they no longer contain <<<WEB-INF/lib/shared-jar-1.0.0.jar>>>,
+ <<<shared-jar-1.0.0.jar>>> and <<<lib/shared-jar-1.0.0.jar>>> files respectively.
+
+ Also, if you inspect the <<<MANIFEST.MF>>> of WAR, SAR and HAR modules of EAR
+ you will notice that the <<<Class-Path>>> entry has been modified or added,
+ if it was missing, and now has a reference to <<<libs/shared-jar-1.0.0.jar>>>.
+
+ If archive representing EAR module has non standard location of libraries then
```
Comment (elharo) [inline src/site/apt/examples/skinny-modules.apt.vm:130]: an archive representing the EAR module has a non-standard location of libraries, then

File: src/site/apt/examples/skinny-modules.apt.vm [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+ this location can be configured using <<<libDirectory>>> property.
```
Comment (elharo) [inline src/site/apt/examples/skinny-modules.apt.vm:131]: using the

File: src/site/apt/index.apt.vm [hunk 15]
```diff
@@ -63,13 +63,18 @@ ${project.name}
   For more information about the Java EE technology, please visit
   {{http://www.oracle.com/us/technologies/java/enterprise-edition/overview/index.html}}.
 
+* Version 3.1.1
+
+  Starting with version 3.1.1 {{{./ear-mojo.html#skinnyModules}skinnyModules}} parameter
```
Comment (elharo) [inline src/site/apt/index.apt.vm:68]: Delete "Starting with version 3.1.1" as it's redundant with the header.

**Comment by mabrarov** [comment]
> https://ci-builds.apache.org/job/Maven/job/maven-box/job/maven-ear-plugin/job/mear-153/

It looks like smth wrong with Jenkins - build looks hanged on checkout phase:

Jenkins build for MEAR-153

File: src/site/apt/index.apt.vm (hunk 15 above)
Comment (mabrarov) [reply src/site/apt/index.apt.vm]: Addressed in 83a46e0

File: src/site/apt/examples/skinny-modules.apt.vm (hunk 14 above)
Comment (mabrarov) [reply src/site/apt/examples/skinny-modules.apt.vm]: Addressed in 83a46e0

File: src/site/apt/examples/skinny-modules.apt.vm (hunk 13 above)
Comment (mabrarov) [reply src/site/apt/examples/skinny-modules.apt.vm]: Addressed in 83a46e0

File: src/site/apt/examples/skinny-modules.apt.vm (hunk 12 above)
Comment (mabrarov) [reply src/site/apt/examples/skinny-modules.apt.vm]: Addressed in 83a46e0

File: src/site/apt/examples/skinny-modules.apt.vm (hunk 11 above)
Comment (mabrarov) [reply src/site/apt/examples/skinny-modules.apt.vm:121]: Addressed in 83a46e0

File: src/main/java/org/apache/maven/plugins/ear/EarMojo.java (hunk 6 above)
Comment (mabrarov) [reply src/main/java/org/apache/maven/plugins/ear/EarMojo.java]: Addressed in 83a46e0

File: src/main/java/org/apache/maven/plugins/ear/EarMojo.java (hunk 5 above)
Comment (mabrarov) [reply src/main/java/org/apache/maven/plugins/ear/EarMojo.java]: Addressed in 83a46e0

File: src/main/java/org/apache/maven/plugins/ear/EarModule.java (hunk 4 above)
Comment (mabrarov) [reply src/main/java/org/apache/maven/plugins/ear/EarModule.java]: Addressed in 83a46e0

File: src/main/java/org/apache/maven/plugins/ear/AbstractEarModule.java (hunk 3 above)
Comment (mabrarov) [reply src/main/java/org/apache/maven/plugins/ear/AbstractEarModule.java]: Addressed in 83a46e0

File: src/main/java/org/apache/maven/plugins/ear/AbstractEarModule.java (hunk 2 above)
Comment (mabrarov) [reply src/main/java/org/apache/maven/plugins/ear/AbstractEarModule.java]: Addressed in 83a46e0

File: src/main/java/org/apache/maven/plugins/ear/EarMojo.java (hunk 7 above)
Comment (mabrarov) [reply src/main/java/org/apache/maven/plugins/ear/EarMojo.java]: Addressed in 83a46e0

File: src/main/java/org/apache/maven/plugins/ear/EarMojo.java (hunk 8 above)
Comment (mabrarov) [reply src/main/java/org/apache/maven/plugins/ear/EarMojo.java]: Addressed in 83a46e0

File: src/site/apt/examples/skinny-modules.apt.vm (hunk 9 above)
Comment (mabrarov) [reply src/site/apt/examples/skinny-modules.apt.vm]: Addressed in 83a46e0

Comment (mabrarov) [reply src/site/apt/examples/skinny-modules.apt.vm]: Addressed in 83a46e0

File: src/test/resources/projects/project-094/ear/pom.xml (hunk 1 above)
Comment (mabrarov) [reply src/test/resources/projects/project-094/ear/pom.xml:49]: Refer to https://github.com/mojohaus/jboss-packaging-maven-plugin/pull/3

**Comment by mabrarov** [comment]
What needs to be done for merging this pull request or what is missing / pending in this pull request? Is there something I have to add to pass review and merge this pull request?

**Comment by elharo** [comment]
At this point, you need to locate a Java EE savvy committer who understands the details of the EAR format and what this plugin is used for to review and approve. I don't use this plugin or Java EE so I can't help with that part.

**Comment by hboutemy** [comment]
sorry, I have been away these times...
FYI, I'm not a EAR user either, then I don't have strong opinion on use cases, sorry. But I want to help people doing hard and good work like this one.
I have 2 questions to evaluate before merging:
1. IIUC, the new feature is triggered by a specific parameter: then can't break existing users configuration?
2. shouldn't we update the version to 3.2.0?

**Comment by mabrarov** [comment]
Hi @hboutemy,

> the new feature is triggered by a specific parameter: then can't break existing users configuration?

Yes. That's one of my intentions - changes should not impact existing projects and the new behavior should work only for users who explicitly want / need it.

> shouldn't we update the version to 3.2.0?

I believe version needs to be changed (3.2.0 looks like a good candidate), but I don't know how it is performed in this project. I thought version change is performed by maintainers when decision on new release is made.

**Comment by hboutemy** [comment]
great: yes, questioning on users@maven.apache.org seems a good option

defining the next version happens when working on an issue where it becomes clear that we're beyond bugfix: yours is like this
Please update your PR with that 3.2.0-SNAPSHOT, I'll do the Jira part

I'm waiting for your feedback to merge this PR, then also to start a release

Thanks a lot, it's great to get such contribution

**Comment by mabrarov** [comment]
@hboutemy 

> then also to start a release

Please take a look at pull request #22 before that.

**Comment by mabrarov** [comment]
@hboutemy,

> Please update your PR with that 3.2.0-SNAPSHOT, I'll do the Jira part

PR updated as requested.

File: src/site/apt/modules.apt.vm [hunk 16]
```diff
@@ -116,6 +116,30 @@ EAR Modules
...
+
+    * Existing elements of <<<Class-Path>>> setting which match libraries packaged
+    in EAR are replaced with references to the matching libraries in EAR.
+
+    * Other libraries packaged in EAR are appended to <<<Class-Path>>> setting in order
+    of EAR dependency on library if
+    {{{./ear-mojo.html#skipClassPathModification}skipClassPathModification}} parameter
+    is false.
+
+  If <<<libDirectory>>> is not <<<null>>> and if
```
Comment (rfscholte) [inline src/site/apt/modules.apt.vm:138]: This makes we wonder if the explicit skinnyModules parameter is required. Setting the libDirectory seems enough to understand that artifacts from the modules are moved here.

**Review by rfscholte — COMMENTED** [review]
Overall a very nice improvement and it matched my idea that a skinnyWar (as defined in the maven-war-plugin) is not the right apporach when working with ears, it is the maven-ear-plugin its responsibiltiy to extract the libs.
One thing we need to keep in mind is that there will be a request to select per module which artifacts should be moved to the (shared) lib directory. Would such a change fit in this PR without adjusting the new parameters? For example, the suggested `skinnyModules` is a boolean (all or nothing), which makes it too hard to finetune such requirements, hence I don't think a global skinnyModules is the right approach here.

File: src/site/apt/modules.apt.vm (hunk 16 above)
Comment (mabrarov) [reply src/site/apt/modules.apt.vm:138]: Sorry, but I didn't get your note. `libDirectory` here is a property of EAR module. If it's used (if it's not `null`, by default or by explicit setting) when `skinnyModules` option is tuned **off** (default value of `skinnyModules` option) then modification of `Class-Path` entry of MANIFEST.MF is **not performed**. This is what this part of documentation tries to say.

**Comment by mabrarov** [comment]
@ rfscholte,

> there will be a request to select per module which artifacts should be moved to the (shared) lib directory.

Do you mean "select skinny mode (move all or nothing) per module and not for all modules"? 

If yes, then here are my thoughts:

I thought about the same but then decided that `skinnyWars` option doesn't provide such ability so that `skinnyModules` option can follow the same approach.

> Would such a change fit in this PR without adjusting the new parameters? 

We can add both:

1. `skinnyModules` option of plugin configuration (implemented)
1. `skinny` property of EAR module which default value is equal to `skinnyModules` option of plugin configuration (`skinnyModules || skinnyWars` for web module).

That way we can use global configuration and perform fine tuning when required. But I'm not sure if this feature is really wanted by users of Maven EAR Plugin... I never had a need:

1. to move into shared EAR lib directory just a part of EAR module libraries
1. to make some modules of EAR skinny while other - non-skinny.

**Comment by rfscholte** [comment]
Just as an example: suppose you have several wars in an EAR and you know that they all use the same logging framework, so these artifacts are the once you want to extract from the WAR and put it in the lib of the EAR. it might also another set of artifacts. The choice of doing a subset and not all might have to do with the classloading strategies.

**Comment by mabrarov** [comment]
> Just as an example: suppose you have several wars in an EAR and you know that they all use the same logging framework, so these artifacts are the once you want to extract from the WAR and put it in the lib of the EAR. it might also another set of artifacts. The choice of doing a subset and not all might have to do with the classloading strategies.

Just don't add non-shared libraries into EAR dependencies - refer to the note at the end of [my previous comment](/apache/maven-ear-plugin/pull/24#issuecomment-727563337).

**Comment by rfscholte** [comment]
> Just don't add non-shared libraries into EAR dependencies
The world isn't that perfect, you can't always control that, especially when working with third party artifacts. There is a reason why this part of the spec isn't implemented for the plugin, it is very difficult to do it right. And Maven is missing some optimizations. Most ear dependencies shouldn't download transitive dependencies. The work around is adding exclusions with wildcards. But that's only an instruction for Maven to download, it is not an instruction for the maven-ear-plugin what not to include.
I'll need some time to see if there is a better way to solve this.

**Comment by mabrarov** [comment]
> But that's only an instruction for Maven to download, it is not an instruction for the maven-ear-plugin what not to include.

I agree, but it's the way Maven EAR Plugin is implemented at the moment. Library is not considered "shared" until it is found among dependencies of EAR. It looks like implementer of skinnyWars option cut corners.

**Comment by mabrarov** [comment]
Just a case / proposal for the further work on design of inclusion / exclusion:

```text
eartest-war-xxx.war
└───WEB-INF/lib/
    ├─── lib1-lib11-xxx.jar
    ├─── lib1-lib12-xxx.jar
    ├─── lib2-lib21-xxx.jar
    ├─── lib2-lib22-xxx.jar
    └─── lib2-lib23-xxx.jar

eartest-rar-xxx.rar
├─── lib1-lib11-xxx.jar
├─── lib1-lib12-xxx.jar
├─── lib2-lib21-xxx.jar
├─── lib2-lib22-xxx.jar
└─── lib2-lib23-xxx.jar
```

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>eartest</groupId>
    <artifactId>ear</artifactId>
    <version>xxx</version>
    <packaging>ear</packaging>
    <dependencies>
        <dependency>
            <groupId>eartest</groupId>
            <artifactId>war</artifactId>
            <version>xxx</version>
            <type>war</type>
        </dependency>
        <dependency>
            <groupId>eartest</groupId>
            <artifactId>war</artifactId>
            <version>xxx</version>
            <type>pom</type>
            <exclusions>
                <!-- Global exclusion for skinnyModules -->
                <exclusion>
                    <groupId>lib2</groupId>
                    <artifactId>lib23</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
        <dependency>
            <groupId>eartest</groupId>
            <artifactId>rar</artifactId>
            <version>xxx</version>
            <type>rar</type>
        </dependency>
        <dependency>
            <groupId>eartest</groupId>
            <artifactId>rar</artifactId>
            <version>xxx</version>
            <type>pom</type>
            <exclusions>
                <!-- Global exclusion for skinnyModules -->
                <exclusion>
                    <groupId>lib2</groupId>
                    <artifactId>lib23</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
    </dependencies>
    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-ear-plugin</artifactId>
                <configuration>
                    <skinnyModules>true</skinnyModules>
                    <modules>
                        <webModule>
                            <groupId>eartest</groupId>
                            <artifactId>war</artifactId>
                            <sharedLibs>
                               <inlcusions>
                                  <inclusion>
                                      <groupId>lib1</groupId>
                                      <artifactId>lib11</artifactId>
                                  </inclusion>
                               </inlcusions>
                            </sharedLibs>
                        </webModule>
                        <rarModule>
                            <groupId>eartest</groupId>
                            <artifactId>rar</artifactId>
                            <sharedLibs>
                               <exclusions>
                                  <exclusion>
                                      <groupId>lib2</groupId>
                                      <artifactId>*</artifactId>
                                  </exclusion>
                               </exclusions>
                            </sharedLibs>
                        </rarModule>
                    </modules>
                </configuration>
            </plugin>
        </plugins>
    </build>
</project>
```

```text
eartest-ear-xxx.ear
├─── lib
│    ├─── lib1-lib11-xxx.jar
│    ├─── lib1-lib12-xxx.jar
│    ├─── lib2-lib21-xxx.jar (Why? Because it's among dependencies of eartest:ear)
│    └─── lib2-lib22-xxx.jar (Why? Because it's among dependencies of eartest:ear)
│
├─── eartest-war-xxx.war
│    └─── WEB-INF/lib/
│         ├─── lib1-lib12-xxx.jar
│         ├─── lib2-lib21-xxx.jar
│         ├─── lib2-lib22-xxx.jar
│         └─── lib2-lib23-xxx.jar
│
└─── eartest-rar-xxx.rar
     ├─── lib2-lib21-xxx.jar
     ├─── lib2-lib22-xxx.jar
     └─── lib2-lib23-xxx.jar
```

**Comment by rfscholte** [comment]
I had a look at https://maven.apache.org/plugins/maven-ear-plugin/examples/skinny-wars.html to understand the current implementation. It is a little bit different to what I was expecting, but it makes sense.
Your last proposal with the sharedLibs per modules combined with includes/excludes is what I was expecting and give maximum control.

**Comment by mabrarov** [comment]
@hboutemy, 

The ["Maven EAR Plugin 3.2.0 with skinnyModules option"](https://lists.apache.org/thread.html/rbcf8c12948893ff80483e2456fb6bdd78a20c94720746a94ad474741%40%3Cusers.maven.apache.org%3E) topic I started in the users@maven.apache.org mailing list has no answers for a week. It looks like we won't get any feedback except [the one](/apache/maven-ear-plugin/pull/24#pullrequestreview-530790385) which we got here from @rfscholte. Maybe it's time to complete with this pull request and pull request #22 (these 2 pull request will produce merge conflicts which I resolved in [merge/MEAR-153_into_MEAR-216](/mabrarov/maven-ear-plugin/tree/merge/MEAR-153_into_MEAR-216) branch) and then release Maven EAR Plugin 3.2.0. Thank you for your help.

@rfscholte,

If you agree with [my proposal](/apache/maven-ear-plugin/pull/24#issuecomment-727590153) and with my suggestion to implement the proposal in a dedicated (new) feature request (pull request), then please confirm that here (approval of this pull request is appreciated too). Please note that this pull request doesn't prevent implementation of that proposal in a backward compatible way - this is the thing which I care about, because I'd prefer to not implement `skinnyModules` feature at all if it will make impossible giving later the maximum control you suggested. Thank you for sharing this idea. I'll open Jira ticket for `sharedLibs` property of EAR module. Maybe I'll be able to work on it this year.

**Comment by mabrarov** [comment]
@hboutemy,

> I'm waiting for your feedback to merge this PR, then also to start a release

Let's start working on this if nobody has concerns.

**Comment by rfscholte** [comment]
Looks good to me

**Comment by mabrarov** [comment]
Hey guy, is it possible to release EAR plugin 3.2.0 with #24 and #22 this year? It would be a nice New Year's gift :) Thank you for your patience and work.

**Comment by rfscholte** [comment]
Let's see what I can do, not sure if I have enough time to do it this year.

**Comment by mabrarov** [comment]
It looks like this pull request was partially merged, because the diff has reduced but is still not empty:

https://github.com/apache/maven-ear-plugin/compare/master...mabrarov:MEAR-153

screenshot1

**Comment by mabrarov** [comment]
I fixed [that issue](https://github.com/apache/maven-ear-plugin/pull/24#issuecomment-739500692) with not empty diff by merge master branch into the source branch of this pull request. No this pull request has empty diff. @hboutemy, it looks like you merged source branch of this pull request into the master branch in 954b7363f7d522c0477207ca16b13b64272650fd, so I request you to take decision on this pull request - whether to merge it for correct history, or decline it (because all changes were merged somehow without this PR).

**Comment by hboutemy** [comment]
Hi Marat,
yes, I merged this PR but amended the last commit message because it was written against MEAR-216 instead of MEAR-153: I was waiting for Apache Jenkins job to finish before closing PR, Jira, and thanking you much for your hard work.

Did I miss something?

Regards,

Hervé

**Comment by mabrarov** [comment]
Hi @hboutemy,

> Did I miss something?

No you didn't. It looks like you merged all changed I proposed in this PR. I'm just confused what to do with this PR.

**Comment by hboutemy** [comment]
It has been merged, even if GitHub can't really get it because of the last commit I amended...
let's close
and more importantly enjoy and say a big thank you for this nice new feature

**Comment by jira-importer** [comment]
Resolve #335

**Comment by jira-importer** [comment]
Resolve #335

**Pull request closed without merging**